### ISSUE TITLE
Backward looking macros: Feature Request 2.0

### ISSUE THREAD

**Issue by pdeffebach** [JuliaLang/julia#47895]
This has been discussed multiple times on Slack and Discord, but I want to file an issue here so it can be formally tracked. 

Something I would really like in Julia is "backwards looking macros", broadly similar to R's `%>%` operator magic. But don't worry, this is *not* yet another piping thread. 

I really like Chain.jl and use it all the time. ie. `@chain x f1(a) f2(b)` becomes `f2(f1(x, a), b)`. The only problem is that it's hard to work with interactively. Often times I write `x` *first* and then only later realize I need to do something to it. In R, a frequent pattern I have is 

```
R> f(a, b, c, df %>% filter(x == 1))
```

where the last `filter` is added because I realized I only wanted to consider observations where `x == 1` in the operation. But I realized this after I had already been typing. 

I recognize there is [InfixFunctions.jl](https://github.com/MasonProtter/InfixFunctions.jl) , but it lacks the same flexibility as metaprogramming. 

I do not know how complicated this would be to implement. But my preferred syntax would be something along the lines of 

```
backwards macro xx(args...)
    ...
end
```

which is *exactly* like a normal macro, but it captures the preceding element. as well. In this way 

```
y @_xx_backwards a b c
```

behaves *exactly the same* as 

```
@xx(y, a, b, c)
```

Thank you!

**Comment by pdeffebach** [comment]
This is not exactly the same though. Here `@f` creates an anonymous function, which has a small overhead in interactive work. It also requires more notation, `|>` and `@f`, so I still think this would be a valuable feature to have.

**Comment by uniment** [comment]
Other than providing a chaining syntax à la magrittr, what benefits would such a syntax bring?

**Comment by pdeffebach** [comment]
Binary operators are nice to type in general I suppose. I'm glad we aren't lisp having to write `+(a, b, -(c, d)))`. It's a nice thing to type.

**Comment by uniment** [comment]
I agree that infix operators are nice. They also necessitate the complexity of operator precedence rules.

I can see how infix macro support is useful for magrittr, but that's only because I'm familiar with the chaining problem. I'm too unimaginative to consider how it could be useful in other contexts though. Do you have any examples?

**Comment by uniment** [comment]
[issue #11608](https://github.com/JuliaLang/julia/issues/11608) is worth reading btw. There was a short window of time when `~` was a special infix macro; the question arose, whether to generalize it or remove it. The latter was chosen.

**Comment by mbauman** [comment]
Surely I must be missing something here — is your preferred calling syntax alternative to `@xx(a, b, c)` **really** `a @_xx_backwards b c`?

**Comment by pdeffebach** [comment]
> Surely I must be missing something here — is your preferred calling syntax alternative to `@xx(a, b, c)` **really** `a @_xx_backwards b c`?

No, definitely not! 

I don't have a concrete syntax proposal. But something like `x |mymacro| expr` might be nice. 

w.r.t. Piping that automatically creates functions. I have a strong preference towards a feature which *only* transforms expressions, without necessarily creating a new scope, since it allows a lot more flexibility, importantly the ability to create variables inside a macro block which are visible outside of it.

**Comment by linwaytin** [comment]
Maybe not directly related to your proposal, but I found [Chevrons](https://github.com/cjdoris/Chevrons.jl) might be an good way to address this problem. The macro it provides can apply recursively, so we can add it at the beginning of a module, and the rules it provides can be used without adding `@bar` at individual lines. 

It also supports REPL.
